### PR TITLE
PXB-2279 Xbcloud: Upload failed: backup is incomplete

### PR DESCRIPTION
**Pull request by rahulmalik87**
PXB-2279 Xbcloud: Upload failed: backup is incomplete

    Problem:
    When backup is a compressed backup, xbcloud "put" operation fails with an error message
    The "put" operation is successful but wrongly throws that it failed

    Analysis:
    To validate that xbcloud put operation is successful, xbcloud looks for the
    presence of ""xtrabackup_info.00000000000000000000" in the list of files
    in cloud storage. In case of compressed backup, the filename is
    "xtrabackup_info.gz.00000000000"

    Because of the extra ".gz" in the filename, it doesn't match with the expected
    name and xbcloud wrongly prints the error message.

    Fix:
    Change the check to real last file "xtrabackup_tablespaces*" instead of "xtrabackupinfo*"
    Look for the prefix "xtrabackup_tablespaces" without the extensions

### CODE REVIEW

**Comment by rahulmalik87** [comment]
Jenkins results https://pxb.cd.percona.com/view/PXB%208.0/job/percona-xtrabackup-8.0-param/lastCompletedBuild/

File: storage/innobase/xtrabackup/src/xbcloud/xbcloud.cc [hunk 1]
```diff
@@ -708,8 +708,17 @@ bool xbcloud_put(Object_store *store, const std::string &container,
...
+    if (cur_file.size() > last_file_expecting.size() &&
+        cur_file.substr(0, last_file_expecting.size())
+                .compare(last_file_expecting) == 0) {
+      file_found = true;
+      break;
+    }
+  }
+  if (!file_found) {
     msg_ts("%s: Upload failed: backup is incomplete.\n", my_progname);
     return false;
```
Comment (satya-bodapati) [inline storage/innobase/xtrabackup/src/xbcloud/xbcloud.cc]: Not clear what you are trying to do? before this change you were looking for the file "xtrabackup_info.00000000000000000000" from the cloud backup dir. What is the new strategy for compressed files? Do they end with .xbcrypt extension? Can you please explain with example?

**Review by satya-bodapati — CHANGES_REQUESTED** [review]

File: storage/innobase/xtrabackup/src/xbcloud/xbcloud.cc (hunk 1 above)
Comment (rahulmalik87) [reply storage/innobase/xtrabackup/src/xbcloud/xbcloud.cc]: they end with xtrabackup_info.gz.00000000000 so now we are just checking if there exists file initial substring is xtrabackup_info. So now xtrabackup_info.gz.00000000000 or xtrabackup_info.00000000000000000000 both will pass the check

Comment (satya-bodapati) [reply storage/innobase/xtrabackup/src/xbcloud/xbcloud.cc]: thanks for explanation

File: storage/innobase/xtrabackup/src/xbcloud/xbcloud.cc [hunk 2]
```diff
@@ -708,8 +708,17 @@ bool xbcloud_put(Object_store *store, const std::string &container,
   }
 
   // check if xtrabackup_info, almost the last file exists in backup
-  if (std::count(object_list.begin(), object_list.end(),
-                 backup_name + "/xtrabackup_info.00000000000000000000") == 0) {
+  std::string last_file_expecting = backup_name + "/xtrabackup_info";
```
Comment (satya-bodapati) [inline storage/innobase/xtrabackup/src/xbcloud/xbcloud.cc]:
```suggestion
  std::string last_file_prefix = backup_name + "/xtrabackup_info";
```

File: storage/innobase/xtrabackup/src/xbcloud/xbcloud.cc [hunk 3]
```diff
@@ -708,8 +708,17 @@ bool xbcloud_put(Object_store *store, const std::string &container,
...
+  bool file_found = false;
+  for (auto cur_file : object_list) {
```
Comment (satya-bodapati) [inline storage/innobase/xtrabackup/src/xbcloud/xbcloud.cc]: if the objec_list maintains insert order, can we check only the last file?
Is there a reverse iterator? ie. from last file in the object_list to the first one?

Comment (satya-bodapati) [inline storage/innobase/xtrabackup/src/xbcloud/xbcloud.cc]: Please add a comment here that we have two possibilities for the last filename  xtrabackup_info.gz.00000000000 or xtrabackup_info.00000000000000000000 and we try to match the prefix "backup/xtrabackup_info" to determine if xbcloud is successful or not

**Review by satya-bodapati — CHANGES_REQUESTED** [review]

File: storage/innobase/xtrabackup/src/xbcloud/xbcloud.cc (hunk 3 above)
Comment (rahulmalik87) [reply storage/innobase/xtrabackup/src/xbcloud/xbcloud.cc]: it is not exactly the last file. the last could be xtrabackup_tablespaces. also, the exact order of xtrabackup_info is not certain. Either we check for xtrabackup_tablespaces. 
I am not sure if we have some option with which we can generate some additional file in end.

Comment (satya-bodapati) [reply storage/innobase/xtrabackup/src/xbcloud/xbcloud.cc]: ok, two questions
1.  Why don't we check for the "real" last file?
2. What about the reverse iterator?

Comment (rahulmalik87) [reply storage/innobase/xtrabackup/src/xbcloud/xbcloud.cc]: fixed, using the real last file. 
tested with --generate-transition-key and --encryption. works fine.

Comment (satya-bodapati) [reply storage/innobase/xtrabackup/src/xbcloud/xbcloud.cc]: great and for question 2?

Comment (rahulmalik87) [reply storage/innobase/xtrabackup/src/xbcloud/xbcloud.cc]: already using reverse iterator

File: storage/innobase/xtrabackup/src/xbcloud/xbcloud.cc [hunk 4]
```diff
@@ -707,9 +707,22 @@ bool xbcloud_put(Object_store *store, const std::string &container,
...
+  /* check if xtrabackup_tablespaces, the last file exists in backup we have two
+    possibilities for the last filename xtrabackup_tablespaces.gz.00000000000 or
+    xtrabackup_tablespaces.00000000000000000000 and we try to match the prefix
+    "backup/xtrabackup_tablespaces" to determine if xbcloud is successful or not
+  */
+  std::string last_file_prefix = backup_name + "/xtrabackup_tablespaces";
+  bool file_found = false;
+
+  if (object_list.size() > 2) {
+    auto cur_file = *(object_list.end() - 1);
```
Comment (satya-bodapati) [inline storage/innobase/xtrabackup/src/xbcloud/xbcloud.cc]: use !vector.empty() and vector.back() instead

**Review by satya-bodapati — CHANGES_REQUESTED** [review]

File: storage/innobase/xtrabackup/src/xbcloud/xbcloud.cc [hunk 5]
```diff
@@ -707,9 +707,22 @@ bool xbcloud_put(Object_store *store, const std::string &container,
...
   }
 
-  // check if xtrabackup_info, almost the last file exists in backup
-  if (std::count(object_list.begin(), object_list.end(),
-                 backup_name + "/xtrabackup_info.00000000000000000000") == 0) {
+  /* check if xtrabackup_tablespaces, the last file exists in backup we have two
+    possibilities for the last filename xtrabackup_tablespaces.gz.00000000000 or
+    xtrabackup_tablespaces.00000000000000000000 and we try to match the prefix
+    "backup/xtrabackup_tablespaces" to determine if xbcloud is successful or not
+  */
```
Comment (satya-bodapati) [inline storage/innobase/xtrabackup/src/xbcloud/xbcloud.cc]: Suggestion:

check if the last_file (xtrabackup_tablespaces.gz.00000000000 or  xtrabackup_tablespaces.00000000000000000000) is uploaded to cloud storage to
determine successful xbcloud "put" operation.

File: storage/innobase/xtrabackup/test/t/xbcloud.sh [hunk 6]
```diff
@@ -139,6 +139,22 @@ if ! grep -q "Upload failed" $topdir/pxb-2202.log ; then
...
 fi
 
+#PXB-2279 Xbcloud: Upload failed: backup is incomplete
+vlog "take full backup to test compression"
+run_cmd xbcloud --defaults-file=$topdir/xbcloud.cnf delete \
+	${full_backup_name} --parallel=4
+rm -r $full_backup_dir
+xtrabackup --backup --stream=xbstream --compress --extra-lsndir=$full_backup_dir \
+	   --target-dir=$full_backup_dir | \
+    run_cmd xbcloud --defaults-file=$topdir/xbcloud.cnf put \
```
Comment (satya-bodapati) [inline storage/innobase/xtrabackup/test/t/xbcloud.sh:149]: can you please add testcase for encrypted backups too?

**Review by satya-bodapati — CHANGES_REQUESTED** [review]
Looks good. Minor comments

Suggestion on Commit message:

Problem:
When backup is a compressed backup, xbcloud "put" operation fails with an error message
The "put" operation is succesful but wrongly throws that it failed

Analysis:
To validate that xbcloud put operation is successful, xbcloud looks for the
presence of ""xtrabackup_info.00000000000000000000" in the list of files
in cloud storage. In case of compressed backup, the filename is
"xtrabackup_info.gz.00000000000"

Because of the extra ".gz" in the filename, it doesn't match with the expected
name and xbcloud wrongly prints the error message.

Fix
----
1. Change the check to real last file "xtrabackup_tablespaces*" instead of "xtrabackupinfo*"
2. Look for the prefix "xtrabackup_tablespaces" without the extensions

**Comment by rahulmalik87** [comment]
$./run.sh -t t/xbcloud.sh
Using /home/rahul.malik/MySQL/src/x8/bld/storage/innobase/xtrabackup/test/var as test root
Detecting server version...
Running against MySQL 8.0.21-debug (InnoDB 8.0.21)
Using 'xtrabackup' as xtrabackup binary
Using 8 parallel workers

==============================================================================
TEST                                   WORKER    RESULT     TIME(s) or COMMENT
------------------------------------------------------------------------------
xbcloud                                  w1	[passed]    106
==============================================================================
Spent 106 of 124 seconds executing testcases

SUMMARY: 1 run, 1 successful, 0 skipped, 0 failed

See results/ for detailed output

**Review by satya-bodapati — APPROVED** [review]

**Pull request merged**